src/Npgsql/Internal/Converters/RangeConverter.cs

```diff
@@ -0,0 +1,189 @@
+
+    async ValueTask<NpgsqlRange<T>> Read(bool async, PgReader reader, CancellationToken cancellationToken)
+    {
+        if (reader.ShouldBuffer(sizeof(byte)))
+            await reader.BufferData(async, sizeof(byte), cancellationToken).ConfigureAwait(false);
+
+        var flags = (RangeFlags)reader.ReadByte();
+        if ((flags & RangeFlags.Empty) != 0)
+            return NpgsqlRange<T>.Empty;
+
+        var lowerBound = default(T);
+        var upperBound = default(T);
+
+        if ((flags & RangeFlags.LowerBoundInfinite) == 0)
+        {
+            if (reader.ShouldBuffer(sizeof(int)))
+                await reader.BufferData(async, sizeof(int), cancellationToken).ConfigureAwait(false);
+            var length = reader.ReadInt32();
+
+            // Note that we leave the CLR default for nulls
+            if (length != -1)
+            {
+                // Set size before calling ShouldBuffer (it needs to be able to resolve an upper bound requirement)
+                reader.Current.Size = length;
+                if (reader.ShouldBuffer(_subtypeBufferReadRequirements))
+                    await reader.BufferData(async, _subtypeBufferReadRequirements, cancellationToken).ConfigureAwait(false);
+
+                lowerBound = async
+                    ? await _subtypeConverter.ReadAsync(reader, cancellationToken)
+                    // ReSharper disable once MethodHasAsyncOverloadWithCancellation
+                    : _subtypeConverter.Read(reader);
+            }
+        }
+
+        if ((flags & RangeFlags.UpperBoundInfinite) == 0)
+        {
+            if (reader.ShouldBuffer(sizeof(int)))
+                await reader.BufferData(async, sizeof(int), cancellationToken).ConfigureAwait(false);
+            var length = reader.ReadInt32();
+
+            // Note that we leave the CLR default for nulls
+            if (length != -1)
+            {
+                // Set size before calling ShouldBuffer (it needs to be able to resolve an upper bound requirement)
+                reader.Current.Size = length;
+                if (reader.ShouldBuffer(_subtypeBufferReadRequirements))
+                    await reader.BufferData(async, _subtypeBufferReadRequirements, cancellationToken).ConfigureAwait(false);
+                upperBound = async
+                    ? await _subtypeConverter.ReadAsync(reader, cancellationToken)
+                    // ReSharper disable once MethodHasAsyncOverloadWithCancellation
+                    : _subtypeConverter.Read(reader);
+            }
+        }
+
+        return new NpgsqlRange<T>(lowerBound, upperBound, flags);
+    }
+
+    public override Size GetSize(SizeContext context, NpgsqlRange<T> value, ref object? writeState)
+    {
+        var totalSize = Size.Create(1); // Flags
+
+        if (!value.IsEmpty)
+        {
+            var rangeWriteState = new WriteState();
+
+            if (!value.LowerBoundInfinite)
+            {
+                var size = Size.Create(4); // Length
```

Comment:
```suggestion
                var size = sizeof(int); // Length
```

There is at least an implicit conversion from int to Size